src/test/ui/layout/unsafe-cell-hides-niche.rs

```diff
@@ -3,30 +3,73 @@
 // test checks that an `Option<UnsafeCell<NonZeroU32>>` has the same
 // size in memory as an `Option<UnsafeCell<u32>>` (namely, 8 bytes).
 
-// run-pass
+// check-pass
+// compile-flags: --crate-type=lib
 
-#![feature(no_niche)]
+#![feature(repr_simd)]
 
-use std::cell::UnsafeCell;
+use std::cell::{UnsafeCell, RefCell, Cell};
 use std::mem::size_of;
 use std::num::NonZeroU32 as N32;
+use std::sync::{Mutex, RwLock};
 
 struct Wrapper<T>(T);
 
 #[repr(transparent)]
 struct Transparent<T>(T);
 
-#[repr(no_niche)]
-struct NoNiche<T>(T);
+struct NoNiche<T>(UnsafeCell<T>);
 
-fn main() {
-    assert_eq!(size_of::<Option<Wrapper<u32>>>(),     8);
-    assert_eq!(size_of::<Option<Wrapper<N32>>>(),     4);
-    assert_eq!(size_of::<Option<Transparent<u32>>>(), 8);
-    assert_eq!(size_of::<Option<Transparent<N32>>>(), 4);
-    assert_eq!(size_of::<Option<NoNiche<u32>>>(),     8);
-    assert_eq!(size_of::<Option<NoNiche<N32>>>(),     8);
+struct Size<const S: usize>;
 
-    assert_eq!(size_of::<Option<UnsafeCell<u32>>>(),  8);
-    assert_eq!(size_of::<Option<UnsafeCell<N32>>>(),  8);
+// Overwriting the runtime assertion and making it a compile-time assertion
+macro_rules! assert_size_eq {
+    ($ty:ty, $size:expr) => {
+        const _: Size::<{$size}> = Size::<{size_of::<$ty>()}>;
+    };
+    ($ty:ty, $size:expr, $optioned_size:expr) => {
+        assert_size_eq!($ty, $size);
+        assert_size_eq!(Option<$ty>, $optioned_size);
+        const _: () = assert!(
+            $size == $optioned_size ||
+            size_of::<$ty>() < size_of::<Option<$ty>>()
```

Comment:
So the reason I brought this up is that this is *the only* check we can use in the cases where you added `if cfg!(target_pointer_width = `.

My understanding is that the size of e.g. `Mutex<()>` varies with *operating system*, not (just) `target_pointer_width`.